File: src/vm.c
#include <math.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "chunk.h"
#include "debug.h"
#include "memory.h"
#include "vm.h"

VM vm;

#define GC_HEAP_GROW_FACTOR 2

/* ---------------------------------------------------------------------
 * Stack helpers
 * ------------------------------------------------------------------- */

void push(Value value) {
    *vm.stackTop = value;
    vm.stackTop++;
}

Value pop(void) {
    vm.stackTop--;
    return *vm.stackTop;
}

Value peek(int distance) {
    return vm.stackTop[-1 - distance];
}

static void resetStack(void) {
    vm.stackTop = vm.stack;
    vm.frameCount = 0;
}

static void runtimeError(const char* format, ...) {
    va_list args;
    va_start(args, format);
    vfprintf(stderr, format, args);
    va_end(args);
    fputc('\n', stderr);

    for (int i = vm.frameCount - 1; i >= 0; i--) {
        CallFrame* frame = &vm.frames[i];
        ObjFunction* fn = frame->function;
        size_t instr = frame->ip - fn->chunk->code - 1;
        int line = fn->chunk->lines[instr];
        fprintf(stderr, "  [line %d] in %s\n", line,
                fn->name ? fn->name->chars : "<script>");
    }
    resetStack();
}

/* ---------------------------------------------------------------------
 * Native functions
 * ------------------------------------------------------------------- */

static Value clockNative(int argCount, Value* args) {
    (void)argCount; (void)args;
    return NUMBER_VAL((double)clock() / CLOCKS_PER_SEC);
}

void defineNative(const char* name, NativeFn function) {
    push(OBJ_VAL(copyString(name, (int)strlen(name))));
    push(OBJ_VAL(newNative(function, name)));
    tableSet(&vm.globals, AS_STRING(vm.stack[0]), vm.stack[1]);
    pop();
    pop();
}

/* ---------------------------------------------------------------------
 * VM lifecycle
 * ------------------------------------------------------------------- */

void initVM(void) {
    resetStack();
    vm.objects = NULL;
    vm.allChunks = NULL;
    vm.bytesAllocated = 0;
    vm.nextGC = 1024 * 1024;
    vm.grayCount = 0;
    vm.grayCapacity = 0;
    vm.grayStack = NULL;

    initTable(&vm.globals);
    initTable(&vm.strings);

    defineNative("clock", clockNative);
}

void freeVM(void) {
    freeTable(&vm.globals);
    freeTable(&vm.strings);
    freeObjects();
}

void registerChunk(Chunk* chunk) {
    chunk->nextChunk = vm.allChunks;
    vm.allChunks = chunk;
}

/* ---------------------------------------------------------------------
 * Garbage collector (mark-sweep). Roots are: the value stack, open call
 * frames' function objects, the globals table, and every constant pool
 * of every chunk ever created (chunks are permanent for the life of the
 * VM, so their constants are always reachable -- mirroring how a JVM's
 * runtime constant pool keeps its entries alive).
 * ------------------------------------------------------------------- */

void markValue(Value value) {
    if (IS_OBJ(value)) markObject(AS_OBJ(value));
}

void markObject(Obj* object) {
    if (object == NULL || object->isMarked) return;
#if GC_LOG
    fprintf(stderr, "%p mark ", (void*)object);
    printValue(OBJ_VAL(object));
    fprintf(stderr, "\n");
#endif
    object->isMarked = true;

    if (vm.grayCapacity < vm.grayCount + 1) {
        vm.grayCapacity = GROW_CAPACITY(vm.grayCapacity);
        vm.grayStack = (Obj**)realloc(vm.grayStack, sizeof(Obj*) * vm.grayCapacity);
        if (vm.grayStack == NULL) { fprintf(stderr, "bcvm: out of memory (gc)\n"); exit(1); }
    }
    vm.grayStack[vm.grayCount++] = object;
}

static void markArray(ValueArray* array) {
    for (int i = 0; i < array->count; i++) markValue(array->values[i]);
}

static void blackenObject(Obj* object) {
    switch (object->type) {
        case OBJ_FUNCTION: {
            ObjFunction* fn = (ObjFunction*)object;
            markObject((Obj*)fn->name);
            markArray(&fn->chunk->constants);
            break;
        }
        case OBJ_NATIVE:
        case OBJ_STRING:
            break; /* no outgoing references */
    }
}

static void markRoots(void) {
    for (Value* slot = vm.stack; slot < vm.stackTop; slot++) markValue(*slot);
    for (int i = 0; i < vm.frameCount; i++) markObject((Obj*)vm.frames[i].function);
    markTable(&vm.globals);

    /* constant pools of every chunk ever compiled are permanent roots */
    for (Chunk* c = vm.allChunks; c != NULL; c = c->nextChunk) {
        markArray(&c->constants);
    }
}

static void traceReferences(void) {
    while (vm.grayCount > 0) {
        Obj* object = vm.grayStack[--vm.grayCount];
        blackenObject(object);
    }
}

static void sweep(void) {
    Obj* previous = NULL;
    Obj* object = vm.objects;
    while (object != NULL) {
        if (object->isMarked) {
            object->isMarked = false;
            previous = object;
            object = object->next;
        } else {
            Obj* unreached = object;
            object = object->next;
            if (previous != NULL) previous->next = object;
            else vm.objects = object;

            switch (unreached->type) {
                case OBJ_STRING: {
                    ObjString* s = (ObjString*)unreached;
                    reallocate(s, sizeof(ObjString) + s->length + 1, 0);
                    break;
                }
                case OBJ_FUNCTION: FREE(ObjFunction, unreached); break;
                case OBJ_NATIVE:   FREE(ObjNative, unreached); break;
            }
        }
    }
}

void collectGarbage(void) {
#if GC_LOG
    fprintf(stderr, "-- gc begin\n");
    size_t before = vm.bytesAllocated;
#endif

    markRoots();
    traceReferences();
    tableRemoveWhite(&vm.strings);
    sweep();

    vm.nextGC = vm.bytesAllocated * GC_HEAP_GROW_FACTOR;

#if GC_LOG
    fprintf(stderr, "-- gc end   collected %zu bytes (%zu -> %zu), next at %zu\n",
            before - vm.bytesAllocated, before, vm.bytesAllocated, vm.nextGC);
#endif
}

/* ---------------------------------------------------------------------
 * Calling convention
 * ------------------------------------------------------------------- */

static bool call(ObjFunction* fn, int argCount) {
    if (argCount != fn->arity) {
        runtimeError("expected %d argument(s) but got %d", fn->arity, argCount);
        return false;
    }
    if (vm.frameCount == FRAMES_MAX) {
        runtimeError("stack overflow");
        return false;
    }
    CallFrame* frame = &vm.frames[vm.frameCount++];
    frame->function = fn;
    frame->ip = fn->chunk->code;
    frame->slots = vm.stackTop - argCount - 1;
    return true;
}

static bool callValue(Value callee, int argCount) {
    if (IS_OBJ(callee)) {
        switch (OBJ_TYPE(callee)) {
            case OBJ_FUNCTION:
                return call(AS_FUNCTION(callee), argCount);
            case OBJ_NATIVE: {
                NativeFn native = AS_NATIVE(callee);
                Value result = native(argCount, vm.stackTop - argCount);
                vm.stackTop -= argCount + 1;
                push(result);
                return true;
            }
            default: break;
        }
    }
    runtimeError("can only call functions (got a %s)", typeName(callee));
    return false;
}

static bool isFalsey(Value v) {
    return IS_NIL(v) || (IS_BOOL(v) && !AS_BOOL(v));
}

static void concatenate(void) {
    ObjString* b = AS_STRING(peek(0));
    ObjString* a = AS_STRING(peek(1));

    int length = a->length + b->length;
    char* chars = ALLOCATE(char, length + 1);
    memcpy(chars, a->chars, a->length);
    memcpy(chars + a->length, b->chars, b->length);
    chars[length] = '\0';

    ObjString* result = takeString(chars, length);
    pop(); pop();
    push(OBJ_VAL(result));
}

/* ---------------------------------------------------------------------
 * The interpreter loop
 * ------------------------------------------------------------------- */

static InterpretResult run(void) {
    CallFrame* frame = &vm.frames[vm.frameCount - 1];

#define READ_BYTE()  (*frame->ip++)
#define READ_SHORT() (frame->ip += 2, (uint16_t)((frame->ip[-2] << 8) | frame->ip[-1]))
#define READ_CONSTANT() (frame->function->chunk->constants.values[READ_BYTE()])
#define READ_STRING() AS_STRING(READ_CONSTANT())

#define BINARY_NUMERIC_OP(valueWrap, op)                                    \
    do {                                                                    \
        if (!IS_NUMBER(peek(0)) || !IS_NUMBER(peek(1))) {                   \
            runtimeError("operands must be numbers");                      \
            return INTERPRET_RUNTIME_ERROR;                                 \
        }                                                                    \
        double b = AS_NUMBER(pop());                                        \
        double a = AS_NUMBER(pop());                                        \
        push(valueWrap(a op b));                                            \
    } while (false)

    for (;;) {
#ifdef BCVM_TRACE
        printf("          ");
        for (Value* slot = vm.stack; slot < vm.stackTop; slot++) {
            printf("[ "); printValue(*slot); printf(" ]");
        }
        printf("\n");
        disassembleInstruction(frame->function->chunk,
                                (int)(frame->ip - frame->function->chunk->code));
#endif
        uint8_t instruction = READ_BYTE();
        switch (instruction) {
            case OP_CONST:  push(READ_CONSTANT()); break;
            case OP_NIL:    push(NIL_VAL); break;
            case OP_TRUE:   push(BOOL_VAL(true)); break;
            case OP_FALSE:  push(BOOL_VAL(false)); break;
            case OP_POP:    pop(); break;
            case OP_DUP:    push(peek(0)); break;

            case OP_GET_LOCAL: push(frame->slots[READ_BYTE()]); break;
            case OP_SET_LOCAL: frame->slots[READ_BYTE()] = peek(0); break;

            case OP_GET_GLOBAL: {
                ObjString* name = READ_STRING();
                Value value;
                if (!tableGet(&vm.globals, name, &value)) {
                    runtimeError("undefined variable '%s'", name->chars);
                    return INTERPRET_RUNTIME_ERROR;
                }
                push(value);
                break;
            }
            case OP_DEFINE_GLOBAL: {
                ObjString* name = READ_STRING();
                tableSet(&vm.globals, name, peek(0));
                pop();
                break;
            }
            case OP_SET_GLOBAL: {
                ObjString* name = READ_STRING();
                if (tableSet(&vm.globals, name, peek(0))) {
                    tableDelete(&vm.globals, name);
                    runtimeError("undefined variable '%s'", name->chars);
                    return INTERPRET_RUNTIME_ERROR;
                }
                break;
            }

            case OP_ADD: {
                if (IS_STRING(peek(0)) && IS_STRING(peek(1))) {
                    concatenate();
                } else if (IS_NUMBER(peek(0)) && IS_NUMBER(peek(1))) {
                    double b = AS_NUMBER(pop());
                    double a = AS_NUMBER(pop());
                    push(NUMBER_VAL(a + b));
                } else {
                    runtimeError("operands must be two numbers or two strings");
                    return INTERPRET_RUNTIME_ERROR;
                }
                break;
            }
            case OP_SUB: BINARY_NUMERIC_OP(NUMBER_VAL, -); break;
            case OP_MUL: BINARY_NUMERIC_OP(NUMBER_VAL, *); break;
            case OP_DIV: BINARY_NUMERIC_OP(NUMBER_VAL, /); break;
            case OP_MOD: {
                if (!IS_NUMBER(peek(0)) || !IS_NUMBER(peek(1))) {
                    runtimeError("operands must be numbers");
                    return INTERPRET_RUNTIME_ERROR;
                }
                double b = AS_NUMBER(pop());
                double a = AS_NUMBER(pop());
                push(NUMBER_VAL(fmod(a, b)));
                break;
            }
            case OP_NEGATE:
                if (!IS_NUMBER(peek(0))) {
                    runtimeError("operand must be a number");
                    return INTERPRET_RUNTIME_ERROR;
                }
                push(NUMBER_VAL(-AS_NUMBER(pop())));
                break;
            case OP_NOT: push(BOOL_VAL(isFalsey(pop()))); break;

            case OP_EQUAL: {
                Value b = pop(), a = pop();
                push(BOOL_VAL(valuesEqual(a, b)));
                break;
            }
            case OP_GREATER: BINARY_NUMERIC_OP(BOOL_VAL, >); break;
            case OP_LESS:    BINARY_NUMERIC_OP(BOOL_VAL, <); break;

            case OP_PRINT: printValue(pop()); printf("\n"); break;

            case OP_JUMP: {
                uint16_t offset = READ_SHORT();
                frame->ip += offset;
                break;
            }
            case OP_JUMP_IF_FALSE: {
                uint16_t offset = READ_SHORT();
                if (isFalsey(peek(0))) frame->ip += offset;
                break;
            }
            case OP_LOOP: {
                uint16_t offset = READ_SHORT();
                frame->ip -= offset;
                break;
            }

            case OP_CALL: {
                int argCount = READ_BYTE();
                if (!callValue(peek(argCount), argCount)) {
                    return INTERPRET_RUNTIME_ERROR;
                }
                frame = &vm.frames[vm.frameCount - 1];
                break;
            }
            case OP_RETURN: {
                Value result = pop();
                vm.frameCount--;
                if (vm.frameCount == 0) {
                    pop(); /* the top-level script "function" value */
                    return INTERPRET_OK;
                }
                vm.stackTop = frame->slots;
                push(result);
                frame = &vm.frames[vm.frameCount - 1];
                break;
            }

            case OP_HALT:
                return INTERPRET_OK;
        }
    }

#undef READ_BYTE
#undef READ_SHORT
#undef READ_CONSTANT
#undef READ_STRING
#undef BINARY_NUMERIC_OP
}

InterpretResult interpret(ObjFunction* entry) {
    push(OBJ_VAL(entry));
    call(entry, 0);
    return run();
}

File: src/vm.rs
//! The virtual machine: value stack, call frames, interpreter loop, and GC.

use std::rc::Rc;

use crate::chunk::OpCode;
use crate::common::{FRAMES_MAX, GC_HEAP_GROW_FACTOR, GC_INITIAL_THRESHOLD, STACK_MAX};
use crate::debug::disassemble_instruction;
use crate::table::Table;
use crate::value::{
    copy_string, new_native, NativeFn, Obj, ObjFunction, ObjRef, Value,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum InterpretResult {
    Ok,
    RuntimeError,
}

/// One activation record on the call stack.
struct CallFrame {
    function: ObjRef,
    /// Index into the function's chunk code (instruction pointer).
    ip: usize,
    /// Base index of this frame's locals inside the shared value stack.
    slots: usize,
}

/// The whole machine state.
pub struct Vm {
    frames: Vec<CallFrame>,
    stack: Vec<Value>,

    pub globals: Table,
    pub strings: Table, // interned string pool

    /// All live heap objects (for the sweep phase).
    objects: Vec<ObjRef>,

    bytes_allocated: usize,
    next_gc: usize,

    /// Gray stack for the mark phase.
    gray_stack: Vec<ObjRef>,

    /// When true, print stack + current instruction each step.
    pub trace_execution: bool,
}

impl Vm {
    pub fn new() -> Self {
        let mut vm = Vm {
            frames: Vec::with_capacity(FRAMES_MAX),
            stack: Vec::with_capacity(STACK_MAX),
            globals: Table::new(),
            strings: Table::new(),
            objects: Vec::new(),
            bytes_allocated: 0,
            next_gc: GC_INITIAL_THRESHOLD,
            gray_stack: Vec::new(),
            trace_execution: false,
        };
        vm.define_native("clock", clock_native);
        vm
    }

    // ------------------------------------------------------------------
    // Stack helpers
    // ------------------------------------------------------------------

    pub fn push(&mut self, value: Value) {
        self.stack.push(value);
    }

    pub fn pop(&mut self) -> Value {
        self.stack.pop().expect("stack underflow")
    }

    pub fn peek(&self, distance: usize) -> &Value {
        &self.stack[self.stack.len() - 1 - distance]
    }

    fn reset_stack(&mut self) {
        self.stack.clear();
        self.frames.clear();
    }

    // ------------------------------------------------------------------
    // Object tracking / GC bookkeeping
    // ------------------------------------------------------------------

    pub fn track_object(&mut self, obj: ObjRef) {
        // Approximate size tracking (good enough for the demo).
        let size = match &*obj.borrow() {
            Obj::String(s) => std::mem::size_of::<Obj>() + s.chars.len(),
            Obj::Function(_) => std::mem::size_of::<ObjFunction>(),
            Obj::Native(_) => std::mem::size_of::<crate::value::ObjNative>(),
        };
        self.bytes_allocated += size;
        self.objects.push(obj);

        #[cfg(feature = "gc_stress")]
        {
            self.collect_garbage();
        }
        #[cfg(not(feature = "gc_stress"))]
        {
            if self.bytes_allocated > self.next_gc {
                self.collect_garbage();
            }
        }
    }

    // ------------------------------------------------------------------
    // Natives
    // ------------------------------------------------------------------

    pub fn define_native(&mut self, name: &'static str, function: NativeFn) {
        let name_obj = copy_string(self, name);
        let native = new_native(self, function, name);
        self.globals.set(name_obj, Value::Obj(native));
    }

    // ------------------------------------------------------------------
    // Runtime error
    // ------------------------------------------------------------------

    fn runtime_error(&mut self, msg: &str) {
        eprintln!("{}", msg);
        for frame in self.frames.iter().rev() {
            let fn_ = frame.function.borrow();
            let f = fn_.as_function();
            let instr = frame.ip.saturating_sub(1);
            let line = f.chunk.lines.get(instr).copied().unwrap_or(0);
            let name = f
                .name
                .as_ref()
                .map(|n| n.borrow().as_string().chars.clone())
                .unwrap_or_else(|| "<script>".to_string());
            eprintln!("  [line {}] in {}", line, name);
        }
        self.reset_stack();
    }

    // ------------------------------------------------------------------
    // Calling convention
    // ------------------------------------------------------------------

    fn call(&mut self, function: ObjRef, arg_count: usize) -> bool {
        let arity = {
            let fn_ = function.borrow();
            fn_.as_function().arity as usize
        };
        if arg_count != arity {
            self.runtime_error(&format!(
                "expected {} argument(s) but got {}",
                arity, arg_count
            ));
            return false;
        }
        if self.frames.len() == FRAMES_MAX {
            self.runtime_error("stack overflow");
            return false;
        }
        let slots = self.stack.len() - arg_count - 1;
        self.frames.push(CallFrame {
            function,
            ip: 0,
            slots,
        });
        true
    }

    fn call_value(&mut self, callee: Value, arg_count: usize) -> bool {
        match callee {
            Value::Obj(ref o) => {
                let kind = {
                    // We need to decide without holding the borrow across the call.
                    let borrowed = o.borrow();
                    match &*borrowed {
                        Obj::Function(_) => 0,
                        Obj::Native(_) => 1,
                        _ => 2,
                    }
                };
                match kind {
                    0 => self.call(Rc::clone(o), arg_count),
                    1 => {
                        let native_fn = {
                            let borrowed = o.borrow();
                            borrowed.as_native().function
                        };
                        let args_start = self.stack.len() - arg_count;
                        let result = native_fn(arg_count, &self.stack[args_start..]);
                        // Pop args + the callee itself.
                        self.stack.truncate(self.stack.len() - arg_count - 1);
                        self.push(result);
                        true
                    }
                    _ => {
                        self.runtime_error(&format!(
                            "can only call functions (got a {})",
                            callee.type_name()
                        ));
                        false
                    }
                }
            }
            _ => {
                self.runtime_error(&format!(
                    "can only call functions (got a {})",
                    callee.type_name()
                ));
                false
            }
        }
    }

    fn is_falsey(v: &Value) -> bool {
        matches!(v, Value::Nil | Value::Bool(false))
    }

    fn concatenate(&mut self) {
        let b = self.peek(0).as_string();
        let a = self.peek(1).as_string();
        let combined = format!("{}{}", a.chars, b.chars);
        // Drop the two operands first so the GC can see the new string as the only root.
        self.pop();
        self.pop();
        let result = copy_string(self, &combined);
        self.push(Value::Obj(result));
    }

    // ------------------------------------------------------------------
    // Interpreter loop
    // ------------------------------------------------------------------

    pub fn interpret(&mut self, entry: ObjRef) -> InterpretResult {
        self.push(Value::Obj(Rc::clone(&entry)));
        if !self.call(entry, 0) {
            return InterpretResult::RuntimeError;
        }
        self.run()
    }

    fn run(&mut self) -> InterpretResult {
        loop {
            let frame_idx = self.frames.len() - 1;

            if self.trace_execution {
                print!("          ");
                for v in &self.stack {
                    print!("[ {} ]", v);
                }
                println!();
                let ip = self.frames[frame_idx].ip;
                // Borrow the function for the duration of disassembly only.
                let fn_ = self.frames[frame_idx].function.borrow();
                disassemble_instruction(&fn_.as_function().chunk, ip);
                drop(fn_);
            }

            // Read the next instruction.
            let instruction = {
                let frame = &mut self.frames[frame_idx];
                let fn_ = frame.function.borrow();
                let chunk = &fn_.as_function().chunk;
                let byte = chunk.code[frame.ip];
                frame.ip += 1;
                OpCode::from(byte)
            };

            match instruction {
                OpCode::Const => {
                    let idx = self.read_byte(frame_idx) as usize;
                    let constant = {
                        let frame = &self.frames[frame_idx];
                        let fn_ = frame.function.borrow();
                        fn_.as_function().chunk.constants.values[idx].clone()
                    };
                    self.push(constant);
                }
                OpCode::Nil => self.push(Value::Nil),
                OpCode::True => self.push(Value::Bool(true)),
                OpCode::False => self.push(Value::Bool(false)),
                OpCode::Pop => {
                    self.pop();
                }
                OpCode::Dup => {
                    let v = self.peek(0).clone();
                    self.push(v);
                }

                OpCode::GetLocal => {
                    let slot = self.read_byte(frame_idx) as usize;
                    let base = self.frames[frame_idx].slots;
                    let v = self.stack[base + slot].clone();
                    self.push(v);
                }
                OpCode::SetLocal => {
                    let slot = self.read_byte(frame_idx) as usize;
                    let base = self.frames[frame_idx].slots;
                    self.stack[base + slot] = self.peek(0).clone();
                }

                OpCode::GetGlobal => {
                    let name = self.read_string(frame_idx);
                    match self.globals.get(&name) {
                        Some(v) => self.push(v),
                        None => {
                            let n = name.borrow().as_string().chars.clone();
                            self.runtime_error(&format!("undefined variable '{}'", n));
                            return InterpretResult::RuntimeError;
                        }
                    }
                }
                OpCode::DefineGlobal => {
                    let name = self.read_string(frame_idx);
                    let value = self.peek(0).clone();
                    self.globals.set(name, value);
                    self.pop();
                }
                OpCode::SetGlobal => {
                    let name = self.read_string(frame_idx);
                    let value = self.peek(0).clone();
                    if self.globals.set(Rc::clone(&name), value) {
                        // set returned true → key was new → undefined
                        self.globals.delete(&name);
                        let n = name.borrow().as_string().chars.clone();
                        self.runtime_error(&format!("undefined variable '{}'", n));
                        return InterpretResult::RuntimeError;
                    }
                }

                OpCode::Add => {
                    if self.peek(0).is_string() && self.peek(1).is_string() {
                        self.concatenate();
                    } else if self.peek(0).is_number() && self.peek(1).is_number() {
                        let b = self.pop().as_number();
                        let a = self.pop().as_number();
                        self.push(Value::Number(a + b));
                    } else {
                        self.runtime_error("operands must be two numbers or two strings");
                        return InterpretResult::RuntimeError;
                    }
                }
                OpCode::Sub => {
                    if let Some(r) = self.binary_numeric(|a, b| a - b) {
                        self.push(Value::Number(r));
                    } else {
                        return InterpretResult::RuntimeError;
                    }
                }
                OpCode::Mul => {
                    if let Some(r) = self.binary_numeric(|a, b| a * b) {
                        self.push(Value::Number(r));
                    } else {
                        return InterpretResult::RuntimeError;
                    }
                }
                OpCode::Div => {
                    if let Some(r) = self.binary_numeric(|a, b| a / b) {
                        self.push(Value::Number(r));
                    } else {
                        return InterpretResult::RuntimeError;
                    }
                }
                OpCode::Mod => {
                    if let Some(r) = self.binary_numeric(|a, b| a % b) {
                        self.push(Value::Number(r));
                    } else {
                        return InterpretResult::RuntimeError;
                    }
                }
                OpCode::Negate => {
                    if !self.peek(0).is_number() {
                        self.runtime_error("operand must be a number");
                        return InterpretResult::RuntimeError;
                    }
                    let n = self.pop().as_number();
                    self.push(Value::Number(-n));
                }
                OpCode::Not => {
                    let v = self.pop();
                    self.push(Value::Bool(Self::is_falsey(&v)));
                }

                OpCode::Equal => {
                    let b = self.pop();
                    let a = self.pop();
                    self.push(Value::Bool(a == b));
                }
                OpCode::Greater => {
                    if let Some(r) = self.binary_numeric(|a, b| a > b) {
                        self.push(Value::Bool(r));
                    } else {
                        return InterpretResult::RuntimeError;
                    }
                }
                OpCode::Less => {
                    if let Some(r) = self.binary_numeric(|a, b| a < b) {
                        self.push(Value::Bool(r));
                    } else {
                        return InterpretResult::RuntimeError;
                    }
                }

                OpCode::Print => {
                    let v = self.pop();
                    println!("{}", v);
                }

                OpCode::Jump => {
                    let offset = self.read_short(frame_idx) as usize;
                    self.frames[frame_idx].ip += offset;
                }
                OpCode::JumpIfFalse => {
                    let offset = self.read_short(frame_idx) as usize;
                    if Self::is_falsey(self.peek(0)) {
                        self.frames[frame_idx].ip += offset;
                    }
                }
                OpCode::Loop => {
                    let offset = self.read_short(frame_idx) as usize;
                    self.frames[frame_idx].ip -= offset;
                }

                OpCode::Call => {
                    let arg_count = self.read_byte(frame_idx) as usize;
                    let callee = self.peek(arg_count).clone();
                    if !self.call_value(callee, arg_count) {
                        return InterpretResult::RuntimeError;
                    }
                    // After a successful call the current frame may have changed.
                }
                OpCode::Return => {
                    let result = self.pop();
                    let frame = self.frames.pop().unwrap();
                    if self.frames.is_empty() {
                        self.pop(); // the top-level script function value
                        return InterpretResult::Ok;
                    }
                    self.stack.truncate(frame.slots);
                    self.push(result);
                }

                OpCode::Halt => return InterpretResult::Ok,
            }
        }
    }

    // ------------------------------------------------------------------
    // Bytecode readers (operate on the current frame)
    // ------------------------------------------------------------------

    fn read_byte(&mut self, frame_idx: usize) -> u8 {
        let frame = &mut self.frames[frame_idx];
        let fn_ = frame.function.borrow();
        let byte = fn_.as_function().chunk.code[frame.ip];
        // We can't mutate ip while the borrow is alive, so drop it first.
        drop(fn_);
        self.frames[frame_idx].ip += 1;
        byte
    }

    fn read_short(&mut self, frame_idx: usize) -> u16 {
        let b1 = self.read_byte(frame_idx) as u16;
        let b2 = self.read_byte(frame_idx) as u16;
        (b1 << 8) | b2
    }

    fn read_string(&mut self, frame_idx: usize) -> ObjRef {
        let idx = self.read_byte(frame_idx) as usize;
        let frame = &self.frames[frame_idx];
        let fn_ = frame.function.borrow();
        let constant = &fn_.as_function().chunk.constants.values[idx];
        match constant {
            Value::Obj(o) => Rc::clone(o),
            _ => panic!("constant is not a string"),
        }
    }

    fn binary_numeric<T, F>(&mut self, op: F) -> Option<T>
    where
        F: FnOnce(f64, f64) -> T,
    {
        if !self.peek(0).is_number() || !self.peek(1).is_number() {
            self.runtime_error("operands must be numbers");
            return None;
        }
        let b = self.pop().as_number();
        let a = self.pop().as_number();
        Some(op(a, b))
    }

    // ------------------------------------------------------------------
    // Mark-sweep garbage collector
    // ------------------------------------------------------------------

    pub fn collect_garbage(&mut self) {
        #[cfg(feature = "gc_log")]
        {
            eprintln!("-- gc begin");
            let before = self.bytes_allocated;
        }

        self.mark_roots();
        self.trace_references();
        self.strings.remove_white();
        self.sweep();

        self.next_gc = self.bytes_allocated * GC_HEAP_GROW_FACTOR;

        #[cfg(feature = "gc_log")]
        {
            eprintln!(
                "-- gc end   collected {} bytes ({} -> {}), next at {}",
                before - self.bytes_allocated,
                before,
                self.bytes_allocated,
                self.next_gc
            );
        }
    }

    fn mark_roots(&mut self) {
        // Collect roots first to avoid simultaneous borrows of `self`.
        // Reachability is: value stack → open call frames → globals table.
        // Function constant pools are traced via blacken when a function is marked.
        let mut roots: Vec<ObjRef> = Vec::new();

        for v in &self.stack {
            if let Value::Obj(o) = v {
                roots.push(Rc::clone(o));
            }
        }
        for frame in &self.frames {
            roots.push(Rc::clone(&frame.function));
        }

        // Globals (keys + values)
        self.globals.mark(&mut |o| roots.push(o));

        for o in roots {
            self.mark_object(o);
        }
    }

    fn mark_object(&mut self, obj: ObjRef) {
        if obj.borrow().is_marked() {
            return;
        }
        #[cfg(feature = "gc_log")]
        {
            eprint!("{:p} mark ", Rc::as_ptr(&obj));
            eprintln!("{}", obj.borrow());
        }
        obj.borrow_mut().set_marked(true);
        self.gray_stack.push(obj);
    }

    fn trace_references(&mut self) {
        while let Some(obj) = self.gray_stack.pop() {
            // Collect outgoing references without holding the borrow.
            let mut children: Vec<ObjRef> = Vec::new();
            {
                let borrowed = obj.borrow();
                match &*borrowed {
                    Obj::Function(f) => {
                        if let Some(ref name) = f.name {
                            children.push(Rc::clone(name));
                        }
                        for v in &f.chunk.constants.values {
                            if let Value::Obj(o) = v {
                                children.push(Rc::clone(o));
                            }
                        }
                    }
                    Obj::String(_) | Obj::Native(_) => {}
                }
            }
            for c in children {
                self.mark_object(c);
            }
        }
    }

    fn sweep(&mut self) {
        let mut i = 0;
        while i < self.objects.len() {
            if self.objects[i].borrow().is_marked() {
                self.objects[i].borrow_mut().set_marked(false);
                i += 1;
            } else {
                // Approximate size accounting.
                let size = match &*self.objects[i].borrow() {
                    Obj::String(s) => std::mem::size_of::<Obj>() + s.chars.len(),
                    Obj::Function(_) => std::mem::size_of::<ObjFunction>(),
                    Obj::Native(_) => std::mem::size_of::<crate::value::ObjNative>(),
                };
                self.bytes_allocated = self.bytes_allocated.saturating_sub(size);
                self.objects.swap_remove(i);
                // Note: the Rc will be dropped when no other references remain.
            }
        }
    }

    pub fn bytes_allocated(&self) -> usize {
        self.bytes_allocated
    }
}

// ---------------------------------------------------------------------------
// Built-in native
// ---------------------------------------------------------------------------

fn clock_native(_arg_count: usize, _args: &[Value]) -> Value {
    // We can't access the Vm's Instant from here easily without making the
    // native signature more complex; a simple process-time approximation
    // is fine for the demo (mirrors the C clock()/CLOCKS_PER_SEC).
    use std::time::{SystemTime, UNIX_EPOCH};
    let secs = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0);
    Value::Number(secs)
}
